sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala

```diff
@@ -125,7 +125,14 @@ object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
     // for DataFrame API cases, same options are carried by both Command and DataSourceV2Relation
     // for DataFrameV2 API cases, options are only carried by Command
     // for SQL cases, options are only carried by DataSourceV2Relation
-    assert(commandOptions == dsOptions || commandOptions.isEmpty || dsOptions.isEmpty)
```

Comment:
The way we turn `CaseInsensitiveStringMap` to a scala Map is `r.options.asScala.toMap`, I think we should change it to `r.options.asCaseSensitiveMap.asScala.toMap` to be case preserving